L1Trigger/L1THGCalUtilities/plugins/CaloTruthCellsProducer.cc

```diff
@@ -0,0 +1,397 @@
+  std::vector<CaloParticleRef> orderedCaloRefs;
+
+  for (unsigned iP(0); iP != caloParticles.size(); ++iP) {
+    auto& caloParticle(caloParticles.at(iP));
+    if (caloParticle.g4Tracks().at(0).eventId().event() != 0) // pileup
+      continue;
+
+    CaloParticleRef ref(caloParticlesHandle, iP);
+
+    SimClusterRefVector const& simClusters(caloParticle.simClusters());
+    for (auto& simCluster : simClusters) {
+      for (auto& hAndF : simCluster->hits_and_fractions()) {
+        DetId hitId(hAndF.first);
+        HGCalDetId tcId;
+        try {
+          tcId = HGCalDetId(geometry.getTriggerCellFromCell(hitId));
+        }
+        catch (cms::Exception const& ex) {
+          edm::LogError("CaloTruthCellsProducer") << ex.what();
+          continue;
+        }
+
+        // tcMap[tcId] += hitMap[hitId.rawId()] * hAndF.first;
+        tcToCalo.emplace(tcId, ref);
+      }
+    }
+
+    // ordered by the gen particle index
+    int genIndex(caloParticle.g4Tracks().at(0).genpartIndex() - 1);
+    if (genIndex < 0) // shouldn't happen
+      continue;
+    if (genIndex >= int(orderedCaloRefs.size()))
+      orderedCaloRefs.resize(genIndex + 1);
+
+    orderedCaloRefs[genIndex] = ref;
+  }
+
+  auto outMap(std::make_unique<CaloToCellsMap>(caloParticlesHandle, triggerCellsHandle));
+  std::unique_ptr<l1t::HGCalTriggerCellBxCollection> outCollection;
+  if (makeCellsCollection_)
+    outCollection.reset(new l1t::HGCalTriggerCellBxCollection);
+
+  int bx(triggerCells.getFirstBX());
+  unsigned bxOffset(0);
+
+  typedef edm::Ptr<l1t::HGCalTriggerCell> TriggerCellPtr;
+  typedef edm::Ptr<l1t::HGCalCluster> ClusterPtr;
+
+  // ClusteringDummyImpl only considers BX 0, so we dump all cells to one vector
+  std::vector<TriggerCellPtr> triggerCellPtrs;
+
+  // loop through all bunch crossings
+  for (unsigned iC(0); iC != triggerCells.size(); ++iC) {
+    if (iC >= bxOffset + triggerCells.size(bx)) {
+      bxOffset += triggerCells.size(bx);
+      bx += 1;
+    }
+
+    auto& cell(triggerCells[iC]);
+
+    auto mapElem(tcToCalo.find(cell.detId()));
+    if (mapElem == tcToCalo.end())
+      continue;
+
+    edm::Ref<l1t::HGCalTriggerCellBxCollection> ref(triggerCellsHandle, iC);
+
+    outMap->insert(mapElem->second, ref);
+
+    if (makeCellsCollection_)
+      outCollection->push_back(bx, cell);
+
+    triggerCellPtrs.emplace_back(triggerCellsHandle, iC);
+  }
+
+  _event.put(std::move(outMap));
+  if (makeCellsCollection_)
+    _event.put(std::move(outCollection), "HGCalVFEProcessorSums");
+
+  auto outClusters(std::make_unique<l1t::HGCalClusterBxCollection>());
+
+  auto sortCellPtrs([](TriggerCellPtr const& lhs, TriggerCellPtr const& rhs)->bool {
+      return lhs->mipPt() > rhs->mipPt();
+    });
+
+  std::sort(triggerCellPtrs.begin(), triggerCellPtrs.end(), sortCellPtrs);
+  dummyClustering_.clusterizeDummy(triggerCellPtrs, *outClusters);
+
+  std::map<CaloParticleRef, std::vector<unsigned>> caloToClusterIndices;
```

Comment:
Would be better to have an `unordered_map` with an integer has key in order to have constant time access

Reply:
Fixed.